File: algorithms/sw_ucb.py
"""
Sliding Window Upper Confidence Bound (SW-UCB) algorithm.
Reference: Garivier & Moulines (2011), arXiv:0805.3415
"""

import math
from collections import deque
from typing import List, Dict, Any
from algorithms.base import BaseAlgorithm, TransactionContext


class SlidingWindowUCB(BaseAlgorithm):
    """
    Non-stationary UCB using a sliding window. Forgets old observations,
    enabling fast adaptation to gateway outages and SR changes.
    """

    def initialize(self, arms: List[str], config: Dict[str, Any]) -> None:
        self.arms = arms
        self.window_size = config.get("window_size", 200)
        self.history: Dict[str, deque] = {
            arm: deque(maxlen=self.window_size) for arm in arms
        }
        self.total_selections = 0
        self._last_scores: Dict[str, float] = {}
        self._last_chosen: str = ""

    def select(self, context: TransactionContext) -> str:
        self.total_selections += 1
        scores = {}

        for arm in self.arms:
            hist = self.history[arm]
            n = len(hist)

            if n == 0:
                self._last_chosen = arm
                self._last_scores = {a: float('inf') if a == arm else 0.0 for a in self.arms}
                return arm

            sr = sum(hist) / n
            exploration_bonus = math.sqrt(2 * math.log(self.total_selections) / n)
            scores[arm] = sr + exploration_bonus

        self._last_scores = scores
        self._last_chosen = max(scores, key=scores.get)
        return self._last_chosen

    def update(self, arm: str, reward: int, context: TransactionContext) -> None:
        self.history[arm].append(reward)

    def get_state(self) -> Dict[str, Any]:
        state = {}
        for arm in self.arms:
            hist = self.history[arm]
            n = len(hist)
            state[arm] = {
                "estimated_sr": sum(hist) / n if n > 0 else None,
                "selection_score": self._last_scores.get(arm),
                "window_count": n,
                "window_capacity": self.window_size,
                "window_successes": sum(hist),
                "window_failures": n - sum(hist),
            }
        return state

    def explain_last_decision(self) -> str:
        if not self._last_chosen or not self._last_scores:
            return "No decision made yet."
        chosen = self._last_chosen
        score = self._last_scores.get(chosen, 0)
        hist = self.history[chosen]
        n = len(hist)
        sr = sum(hist) / n if n > 0 else 0
        bonus = score - sr if score != float('inf') else float('inf')
        others = {a: f"{s:.4f}" for a, s in self._last_scores.items() if a != chosen}
        return (
            f"Chose '{chosen}' with UCB={score:.4f} "
            f"(SR={sr:.3f} + bonus={bonus:.4f}, window_n={n}). "
            f"Other scores: {others}"
        )

    def get_hyperparameter_schema(self) -> Dict[str, Any]:
        return {
            "window_size": {
                "type": "integer",
                "default": 200,
                "min": 10,
                "max": 10000,
                "description": "Number of most recent transactions per gateway to consider. Smaller = faster adaptation but noisier.",
            }
        }

    @classmethod
    def metadata(cls) -> Dict[str, str]:
        return {
            "name": "Sliding Window UCB",
            "short_name": "SW-UCB",
            "description": "Non-stationary UCB using a sliding window. Forgets old observations, enabling fast adaptation to gateway outages and SR changes.",
            "paper": "Garivier & Moulines (2011). On Upper-Confidence Bound Policies for Non-Stationary Bandit Problems. ALT 2011.",
            "paper_url": "https://arxiv.org/abs/0805.3415",
            "category": "bandit",
            "non_stationary": "true",
        }

File: frontend/src/features/dynamic-simulation/algorithms/sw_ucb.ts
/**
 * Sliding Window UCB (SW-UCB) — TypeScript port
 * Reference: Garivier & Moulines (2011), arXiv:0805.3415
 */

import { BaseAlgorithm, SeededRNG } from './base';
import { TransactionContext, ArmState, AlgorithmMetadata, HyperparameterSchema } from '../types';

export class SlidingWindowUCB implements BaseAlgorithm {
    private arms: string[] = [];
    private windowSize = 200;
    private history: Map<string, number[]> = new Map();
    private totalSelections = 0;
    private lastScores: Record<string, number> = {};
    private lastChosen = '';

    initialize(arms: string[], config: Record<string, number>): void {
        this.arms = arms;
        this.windowSize = config.window_size ?? 200;
        this.history = new Map();
        for (const arm of arms) {
            this.history.set(arm, []);
        }
        this.totalSelections = 0;
        this.lastScores = {};
        this.lastChosen = '';
    }

    select(_context: TransactionContext): string {
        this.totalSelections++;
        const scores: Record<string, number> = {};

        for (const arm of this.arms) {
            const hist = this.history.get(arm)!;
            const n = hist.length;
            if (n === 0) {
                this.lastChosen = arm;
                this.lastScores = Object.fromEntries(
                    this.arms.map(a => [a, a === arm ? Infinity : 0])
                );
                return arm;
            }
            const sr = hist.reduce((s, v) => s + v, 0) / n;
            const bonus = Math.sqrt(2 * Math.log(this.totalSelections) / n);
            scores[arm] = sr + bonus;
        }

        this.lastScores = scores;
        this.lastChosen = Object.entries(scores).reduce((best, [arm, score]) =>
            score > best[1] ? [arm, score] : best, ['', -Infinity]
        )[0] as string;
        return this.lastChosen;
    }

    update(arm: string, reward: number, _context: TransactionContext): void {
        const hist = this.history.get(arm)!;
        hist.push(reward);
        if (hist.length > this.windowSize) {
            hist.shift();
        }
    }

    getState(): Record<string, ArmState> {
        const state: Record<string, ArmState> = {};
        for (const arm of this.arms) {
            const hist = this.history.get(arm)!;
            const n = hist.length;
            const sum = hist.reduce((s, v) => s + v, 0);
            state[arm] = {
                estimatedSR: n > 0 ? sum / n : null,
                selectionScore: this.lastScores[arm] ?? null,
                totalSelections: n,
                windowCapacity: this.windowSize,
            };
        }
        return state;
    }

    explainLastDecision(): string {
        if (!this.lastChosen) return 'No decision made yet.';
        const score = this.lastScores[this.lastChosen] ?? 0;
        const hist = this.history.get(this.lastChosen)!;
        const n = hist.length;
        const sr = n > 0 ? hist.reduce((s, v) => s + v, 0) / n : 0;
        return `Chose '${this.lastChosen}' with UCB=${score.toFixed(4)} (SR=${sr.toFixed(3)}, window_n=${n})`;
    }

    getHyperparameterSchema(): HyperparameterSchema {
        return {
            window_size: {
                type: 'integer',
                default: 200,
                min: 10,
                max: 10000,
                description: 'Number of most recent transactions per gateway. Smaller = faster adaptation but noisier.',
            },
        };
    }

    metadata(): AlgorithmMetadata {
        return {
            name: 'Sliding Window UCB',
            shortName: 'SW-UCB',
            description: 'Non-stationary UCB using a sliding window. Forgets old observations for fast adaptation.',
            paper: 'Garivier & Moulines (2011)',
            paperUrl: 'https://arxiv.org/abs/0805.3415',
            category: 'bandit',
        };
    }
}
